sys/src/9/amd64/devacpi.c

```diff
@@ -2066,6 +2093,189 @@ acpimemread(Chan *c, void *a, int32_t n, int64_t off)
+ * to interrupts even though the real interrupt handler runs in the kernel.
+ * XXX need to support pm1b and xpm1 registers
+ */
+static int
+acpiio(uint32_t start, uint32_t end)
+{
+	Acpiio *io;
+
+	for (io = acpiiospace; io < acpiiospace + NACPIPORTS; io++) {
+		if (start >= io->start && start < io->end
+		&& end > start && end <= io->end)
+			return 1;
+	}
+	return 0;
+}
+
+static int32_t
+acpiioread(Chan *c, void *a, int32_t n, int64_t off)
+{
+	int port, pm1aevtblk;
+	uint8_t *p;
+	uint16_t *sp;
+	uint32_t *lp;
+
+	pm1aevtblk = (int)fadt->pm1aevtblk;
+	port = off;
+	if (!acpiio(port, port + n)) {
+		error(Eperm);
+		return -1;
+	}
+	switch (n) {
+	case 1:
+		p = a;
+		if (port == pm1aevtblk)
+			*p = pm1status & 0xff;
+		else if (port == (pm1aevtblk + 1))
+			*p = pm1status >> 8;
+		else
+			*p = inb(port);
+		return n;
+	case 2:
+		if (n & 1) {
+			error(Ebadarg);
+			break;
+		}
+		sp = a;
+		if (port == pm1aevtblk)
+			*sp = pm1status;
+		else
+			*sp = ins(port);
+		return n;
+	case 4:
+		if (n & 3) {
+			error(Ebadarg);
+			break;
+		}
+		lp = a;
+		*lp = inl(port);
+		if (port == pm1aevtblk)
+			*lp |= pm1status;
+		return n;
+	default:
+		error(Ebadarg);
+	}
+	return -1;
+}
+
+static int32_t
+acpiiowrite(Chan *c, void *a, int32_t n, int64_t off)
+{
+	int port, pm1aevtblk;
+	uint8_t *p;
+	uint16_t *sp;
+	uint32_t *lp;
+
+	port = off;
+	pm1aevtblk = (int)fadt->pm1aevtblk;
+	if (!acpiio(port, port + n)) {
+		error(Eperm);
+		return -1;
+	}
+	switch (n) {
+	case 1:
+		p = a;
+		if (port == pm1aevtblk)
+			pm1status &= ~*p;
+		else if (port == (pm1aevtblk + 1))
+			pm1status &= ~(*p << 8);
+		else
+			outb(port, *p);
+		return n;
+	case 2:
+		if (n & 1)
+			error(Ebadarg);
+		sp = a;
+		if (port == pm1aevtblk)
+			pm1status &= ~*sp;
+		outs(port, *sp);
+		return n;
+	case 4:
+		if (n & 3)
+			error(Ebadarg);
+		lp = a;
+		if (port == pm1aevtblk)
+			pm1status &= ~*lp;
+		outl(port, (*lp & ~0xffff));
+		return n;
+	default:
+		error(Ebadarg);
+	}
+	return -1;
+}
+
+static int32_t
+acpiintrread(Chan *c, void *a, int32_t n, int64_t off)
+{
+	if (a == nil)
+		return 0;
+	if (acpiintrvec)
+		return readstr(off, a, n, "1");
+	else
+		return readstr(off, a, n, "0");
+}
+
+static int32_t
+acpiintrwrite(Chan *c, void *a, int32_t n, int64_t off)
+{
+	char *p;
+
+	p = a;
+	if (p == nil || n < 1)
+		return 0;
+	if (*p == '1') {
+		acpiintrvec = intrenable(fadt->sciint, acpihandler, nil, BUSUNKNOWN,
+		    "ACPI interrupt handler");
+		if (acpiintrvec) {
+			intrproc = externup();
+			intrprocpid = intrproc->pid;
+			return 1;
+		} else
+			return 0;
+	} else if (*p == '0') {
+		if (acpiintrvec) {
+			intrdisable(acpiintrvec);
+			acpiintrvec = nil;
+		}
+		intrproc = nil;
+		return 1;
+	} else
+		return 0;
+}
+
+static void
+acpihandler(Ureg *u, void *v)
+{
+	unsigned int status;
+
+	status = getpm1sts();
+	pm1status = status;
+	setpm1sts(status);
+	if (intrproc == nil)
+		return;
+	/* XXX there must be a better way to test if user Proc is okay */
```

Comment:
oh and when the Chan is closed, intrproc is cleared. No locks needed for setting and clearing it, as the Chan and exclusive open ensure that you're the only person manipulating those variables!